modules/boulder/src/main/ts/ephox/boulder/core/ValuePresence.ts

```diff
@@ -0,0 +1,53 @@
+import * as FieldPresence from '../api/FieldPresence';
+import { Processor } from './ValueProcessor';
+
+export const enum ValueType {
+  Field = 'field',
+  State = 'state'
+}
+
+interface FieldData {
+  readonly key: string;
+  readonly newKey: string;
+  readonly presence: FieldPresence.FieldPresenceTypes;
+  readonly prop: Processor;
+}
+
+interface StateData {
+  readonly newKey: string;
+  readonly instantiator: (obj: any) => unknown;
+}
+
+interface ValuePresenceData<D extends ValueType, T> {
+  readonly tag: D;
+  readonly data: T;
+}
+
+export type FieldProcessorData = ValuePresenceData<ValueType.Field, FieldData>;
+export type StateProcessorData = ValuePresenceData<ValueType.State, StateData>;
```

Comment:
Is there a reason we couldn't do this? It leads to shorter error messages if you get the types wrong elsewhere, and closer to how they recommend we make ADTs in the official documentation

```suggestion
export interface FieldProcessorData {
  readonly tag: ValueType.Field;
  readonly key: string;
  readonly newKey: string;
  readonly presence: FieldPresence.FieldPresenceTypes;
  readonly prop: Processor;
}

export interface StateProcessorData {
  readonly tag: ValueType.State;
  readonly newKey: string;
  readonly instantiator: (obj: any) => unknown;
}
```

Reply:
I hadn't considered this, but I believe this should be viable. It'll work as the union type will just have common fields and then when doing a switch statement you'll be able to easily access the properties instead of having to go via `data`.

Reply:
Ahh, I meant to do this after changing the other one. Thanks!